9138 hostname caching policy

jlitzingerdev opened the pull request:
PR for https://twistedmatrix.com/trac/ticket/9138

jlitzingerdev committed:
Add hostname caching policy to twisted.web.client

The HostnameCachingHTTPSPolicy is an IPolicyForHTTPS that caches
connection creators for reuse in subsequent requests.  The cache size is
determined by the cacheSize parameter and defaults to 20.  When a cache
miss occurs, an entry for the new host is created and added to the
cache.  The cache is then purged and the oldest entry removed.

jlitzingerdev committed:
Move CustomOpenSSLTrustRoot out of test function

The custom trust root is useful when testing custom HTTPS policies, move
it outside its current test function if the SSL dependencies are
successfully imported.

hawkowl commented:
@jlitzingerdev Python 3 seems to be failing :(

jlitzingerdev commented:
Uy, Indeed it is, sorry about.  My process was broken, I didn't test python 3 with TLS dependencies installed...process fixed.  I'm assuming I should close, rebase, and submit a new PR, correct?  Do I need to version (thinking of patch versions for kernel submissions) the PR or is a new commit sufficient?

hawkowl commented:
@jlitzingerdev Nah, just go ahead and push another commit to the branch/PR, and it'll update/rerun tests.

jlitzingerdev committed:
Remove a useless line of code.

The entry.creator was cruft from a refactoring that never should have
remained in the branch, remove.

jlitzingerdev committed:
Increment _nextCacheId in except and else instead of finally.

The use of finally in the previous code hid two bugs:

1. If an exception was thrown by optionsForClientTLS, entry would not
exist and an UnboundLocalError would be thrown in the finally clause,
masking the original exception.

2.  If 1 occured, _nextCacheId would still be incremented, leaving a
hole and resulting in no cache entry removed during a later purge (due
to the fact that no entry existed with that value).

jlitzingerdev committed:
Fix Python3 bytes/str issues.

Caught during test-build of pull request.  Unicode objects don't have a
decode method in Python3.  Modify the code such that, when an argument
is passed to creatorForNetloc it is always an ascii encoded byte string,
but when the _cache is manually inspected a unicode (str) object is
used.

jlitzingerdev commented:
@hawkowl Pushed, though it ended up being three commits as there were three distinct issues I found with the previous submission.  I did not rebase on trunk, but will be happy to do so if needed.

jlitzingerdev commented:
@hawkowl I must admin I'm a bit confused on the travis failure.  Most of the jobs passed, and the one that failed "ERROR:   manifest-checker: commands failed", is this related to my branch being out of date with master?

rodrigc commented:
You should merge trunk into your branch to update it,
and also make sure that you have a **news fragment**, as specified in step 6 here:

https://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch

Your lack of a **news fragment** is causing the Travis failure.

jlitzingerdev commented:
@rodrigc I did indeed read the newsifle requirement, and I added a one, see https://github.com/twisted/twisted/pull/793/files#diff-3d267930e0599d705b7ed77fd7572f25, though I added it in the newsfragments directory of the twisted.web package, was that incorrect?

jlitzingerdev commented:
@rodrigc Also, copy that on the merge of trunk, but a question, do you really want a merge, or should I rebase my changes on current trunk (so this PR is always a fast forward)?  My only concern with a merge is that, assuming I have more feedback to incorporate, new changes for this changeset will be spread across a merge commit (somewhat pointlessly).  Ultimately I'll do whatever is preferred, just want the history to look how want it.

rodrigc commented:
Merge is fine.

jlitzingerdev commented:
@rodrigc thanks, one last question pre-push.  Did you see my comment about the newsfile?  I did add one, and the build did pick it up.  It appears the build is complaining about
missing from sdist:
  src/twisted/names/topfiles

Which I'm assuming will be resolved with the merge.  Was there an issue with my newsfragment as submitted?

rodrigc commented:
Yes, I saw the newsfragment you submitted.  Please do the merge.

rodrigc commented:
Your branch has a file src/twisted/names/topfiles/8340.bug fix but that file does not exist on trunk.  Maybe there was a conflict during a merge and you didn't resolve it correctly?

rodrigc commented:
You need to update your tree at https:://github.com/jlitzingerdev/twisted so that it's trunk has all the commits from https://github.com/twisted/twisted .  You are missing commits done later this week which are needed.

Then you need to do another merge from trunk to this branch.

jlitzingerdev commented:
@rodrigc That makes more sense.  I haven't kept my fork's trunk up to date (it is locally, but I didn't push), I was (wrongly) assuming it didn't matter as the PR was against twisted's.  Sorry, one moment.

jlitzingerdev commented:
@rodrigc Done, fork trunk updated, sorry about that.  Re. your other comment, that 8430 does appear to exist in twisted's trunk, but it did not exist in mine at the time.

Can I assume that the tests, when they need trunk use the fork's trunk?  If so, would that be helpful to add to the documentation (unless I just missed it)?  If so I can add it.

Also, I did merge my fork's trunk into this branch, but it was already up to date a I merged trunk from the upstream earlier (git merge upstream/trunk).

rodrigc commented:
@markrwilliams you introduced src/twisted/names/topfiles/8340.bugfix, which is now in an incorrect directory.  This is now causing other pull requests to fail.  Can you fix this?

markrwilliams commented:
@rodrigc I can fix it in this case, but @hawkowl wrote a very handy script: [`admin/fix-for-towncrier.py`](https://github.com/twisted/twisted/blob/trunk/admin/fix-for-towncrier.py).  It recursively renames `topfiles` to `newsfragements`.  Contributors should run this from the top level of the git repository and commit the result.

markrwilliams commented:
I've [updated the wiki](https://twistedmatrix.com/trac/wiki/ReviewProcess?action=diff&version=81)

rodrigc reviewed (commented):
@markrwilliams Can you cherry pick https://github.com/twisted/twisted/commit/98d3e64 to trunk?   The travis trunk build is red without this fix:  https://travis-ci.org/twisted/twisted/branches

jlitzingerdev commented:
@rodrigc @hawkowl @markrwilliams It looks like the latest merge (https://github.com/twisted/twisted/commit/fedec1a2319df1d1ef3d88835d5cc7719a1c9311) introduced the same topfile fix.  Do you want me to merge trunk again, or create a clean v2 branch/PR with the commits pertinent to ticket 9138 (perhaps squashed back down to two commits)?

jlitzingerdev commented:
I rebased this branch against trunk today.  I'd merged trunk in on Sunday, but it was becoming fairly messy.  I assumed rebasing this was desirable per item 15 in https://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch.  That unfortunately required a force-push, which I typically avoid, apologies for any issues that causes.  Note that I also updated my fork's trunk just after this push, which might trigger a CI failure (it seemed to before).  Unless I hear otherwise I'll plan to continue rebasing this against trunk.

jlitzingerdev committed:
Fix test names to match convention.

The existing test names do not follow the twisted naming convention of
test_<testName>, update all tests for the hostname caching policy to
follow the convention.

jlitzingerdev committed:
Merge branch 'trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge branch 'trunk' into 9138-hostname-caching-policy

markrwilliams committed:
Merge branch 'trunk' into 9138-hostname-caching-policy

markrwilliams committed:
Make an LRU out of an OrderedDict.

markrwilliams commented on src/twisted/web/client.py:
```diff
@@ -955,6 +955,112 @@ def creatorForNetloc(self, hostname, port):
+    creator and a cacheId.
+
+    @ivar creator: See L{__init__}
+    @ivar cacheId: See L{__init__}
```
`cacheID`, not `cacheId`.

jlitzingerdev replied:
Thanks, I'll update all these accordingly and be sure to follow this in the future.

markrwilliams commented on src/twisted/web/client.py:
```diff
@@ -955,6 +955,112 @@ def creatorForNetloc(self, hostname, port):
+        @type creator: L{client connection creator
+            <twisted.internet.interfaces.IOpenSSLClientConnectionCreator>}
+
+        @param cacheId: The id of this entry in the cache
```
`cacheID`, not `cacheId`.

markrwilliams commented on src/twisted/web/client.py:
```diff
@@ -955,6 +955,112 @@ def creatorForNetloc(self, hostname, port):
+        @type cacheId: L{int}
+        """
+        self.creator = creator
+        self.cacheId = cacheId
```
`cacheID`, not `cacheId`

markrwilliams commented on src/twisted/web/client.py:
```diff
@@ -955,6 +955,112 @@ def creatorForNetloc(self, hostname, port):
+    @ivar _cacheSize: See C{cacheSize} parameter of L{__init__}.
+    @ivar _nextCacheId: The id to use for the next cache modification,
+        either a hit or a new entry.
+    @type _nextCacheId: L{int}
```
`_nextCacheID`, not `_nextCacheId`.

markrwilliams commented on src/twisted/web/client.py:
```diff
@@ -955,6 +955,112 @@ def creatorForNetloc(self, hostname, port):
+        self._trustRoot = trustRoot
+        self._cache = {}
+        self._cacheSize = cacheSize
+        self._nextCacheId = 0
```
`_nextCacheId`, not `_nextCacheID`.

markrwilliams commented on src/twisted/web/client.py:
```diff
@@ -955,6 +955,112 @@ def creatorForNetloc(self, hostname, port):
+        """
+        Purge cache entries if cacheSize is exceeded.
+        """
+        idToRemove = self._nextCacheId - self._cacheSize
```
There's a bug here :frowning_face:  Because [adding item to the cache also increments `_nextCacheId`](https://github.com/twisted/twisted/pull/793/files#diff-83c0fea4716a44bf877b0b63efb6006aR1060), the distance between the most recently accessed and its successor grows with each access.  After `_cacheSize` accesses, `idToRemove` is larger than the largest ID.

Suppose I create a cache with 3 items:
```python
cache = HostnameCachingHTTPSPolicy(trustRoot="ignored", cacheSize=3)
cache.creatorForNetloc(b"host1", port=80)
cache.creatorForNetloc(b"host2", port=80)
cache.creatorForNetloc(b"host3", port=80)
```
Then, I ask for `b"host1"` 3 times in a row:
```python
for _ in range(3):
    cache.creatorForNetloc(b"host1", port=80)
```
I get an exception when I attempt to get a creator for a new host:

```python
cache.creatorForNetloc(b'host4', port=80)
# Traceback (most recent call last):
# File "bug.py", line 24, in <module>
#    cache.creatorForNetloc(b'host4', port=80)
#  File "/home/mrw/src/python/twisted/src/twisted/web/client.py", line 1055, in creatorForNetloc
#    self._purgeCache()
#  File "/home/mrw/src/python/twisted/src/twisted/web/client.py", line 1028, in _purgeCache
#    del self._cache[keyToRemove]
# UnboundLocalError: local variable 'keyToRemove' referenced before assignment
```

That's because `_nextCacheId` is 6 and the cache size is 3, so `idToRemove` is 3, but the largest cache ID any entry has is 2.

A solution would be to sort the cache keys in increasing order, then remove all but the last `self._cacheSize` items:

```python
def _purgeCache(self):
    """
    Purge cache entries if cacheSize is exceeded.
    """
    items = self._cache.items()
    items.sort(key=lambda kv: kv[1].cacheId)
    del items[:-self._cacheSize]
    self._cache = dict(items)
```

However, I recommend [using `OrderedDict` instead](https://github.com/jlitzingerdev/twisted/pull/1).

jlitzingerdev replied:
Thanks, that's a good find, sorry about the bug.  That use of OrderedDict is fantastic!  I honestly did not even consider it because I didn't realize popitem had those semantics.  I'm more familiar with deque and namedtuple from collections...seems there's always something new to learn in the library.

markrwilliams commented on src/twisted/web/test/test_agent.py:
```diff
@@ -3051,3 +3056,108 @@ def test_deprecatedTransportNoWarning(self):
+            hostname = u"host" + unicode(i)
+            policy.creatorForNetloc(hostname.encode("ascii"), 8675)
+
+        # Force a, which was the first, to be the most recently used
```
Should this read `host0` instead of `a`?

markrwilliams committed:
Assert that many accesses does not violate LRU invariants.

markrwilliams reviewed (dismissed):
This is a fantastic feature, but there's a bug in the implementation.  The review includes two approaches to fixing it.

Thanks for all your work on this!

jlitzingerdev committed:
Merge pull request #1 from markrwilliams/9138-hostname-caching-policy-ordereddict

Make an LRU out of an OrderedDict.

jlitzingerdev committed:
creatorForNetLoc requires a bytes argument.

The existing code passed a str object in python 3 which has no decode
method.  Always ensure the hostnames in the test are bytes.

exarkun commented on src/twisted/web/client.py:
```diff
@@ -955,6 +956,65 @@ def creatorForNetloc(self, hostname, port):
+        try:
+            creator = self._cache.pop(host)
+        except KeyError:
+            creator = optionsForClientTLS(host, trustRoot=self._trustRoot)
```
Instead of duplicating the logic in `BrowserLikePolicyForHTTPS` it would be nice to make this implementation a wrapper around an arbitrary other implementation.  This will also simplify testing.

jlitzingerdev replied:
Seems reasonable and not too difficult.  I'll make the changes and push here to make sure I'm on the right track (I find code makes any subsequent discussions easier) before putting it back into review.

jlitzingerdev replied:
@exarkun Is this along the lines of what you had in mind?
https://github.com/jlitzingerdev/twisted/blob/d48f24ec36683c546486d5878216427a25c654a3/src/twisted/web/client.py#L1010-1021

```
        host = hostname.decode("ascii")
        try:
            creator = self._cache.pop(host)
        except KeyError:
            creator = self._policyForHTTPS.creatorForNetloc(hostname, port)


        self._cache[host] = creator
        if len(self._cache) > self._cacheSize:
            self._cache.popitem(last=False)


        return creator
```

jlitzingerdev committed:
Wrap an IPolicyForHTTPS instead of duplicating behavior.

Rather than duplicate the existing behavior of
BrowserLikePolicyForHTTPS, allow clients to pass in any arbitrary
IPolicyForHTTPS and cache the returned connection creator.  This lets
the various existing (and future) policies manage the creation of the
connector in whatever way suits there needs, but allows for reuse of the
connector for the same hostname.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Improve docstrings.

Refer to the actual relevant type and variable name rather than less
than direct descriptions.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Fix twistedchecker warnings.

Test classes should end with Tests and a python code style bug.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

twm committed:
Merge branch 'trunk' into 9138-hostname-caching-policy

twm commented on src/twisted/web/client.py:
```diff
@@ -956,6 +957,70 @@ def creatorForNetloc(self, hostname, port):
+    L{client connection creators <twisted.internet.interfaces.
+    IOpenSSLClientConnectionCreator>} for reuse in subsequent requests to
+    the same hostname.
+
```
This should have an `@since` annotation as this is a new API. For example:

```
@since: Twisted NEXT
```

The "NEXT" will be replaced with the release number when a release is cut.

wiml replied:
Good point, added that.

hawkowl committed:
Merge branch 'trunk' into 9138-hostname-caching-policy

glyph committed:
Merge branch 'trunk' into 9138-hostname-caching-policy

glyph commented:
This lint failure looks silly; don't have time to do a full review right now, but a committer should probably just come along and pacify the builder and land this.

jlitzingerdev commented:
@glyph I assumed from the age that there wasn't much interest in this enhancement.  I haven't done much in the way of maintenance (as evidenced in the logs above) in quite some time.  I'll clean up the errors and get it in shape for a proper review.

Note that I did submit a PR for the benchmarks (what's the point in an improvement if there is no measurement) here:
https://github.com/twisted-infra/twisted-benchmarks/pull/5
with the intent being to update the benchmark using the code from this PR after it was merged.

jlitzingerdev committed:
Merge remote-tracking branch 'upstream/trunk' into 9138-hostname-caching-policy

jlitzingerdev committed:
Add blank line to comply with Twisted code style.

Caught by checker in review.

jlitzingerdev committed:
Remove blank line to comply with Twisted code style.

Caught during review by twisted checker.

wiml commented:
As an aside, on Python 3, this could also be done using the [`@functools.lru_cache`](https://docs.python.org/3/library/functools.html#functools.lru_cache) decorator.

wiml committed:
Add a @since field to HostnameCachingHTTPSPolicy's docstring.

wiml committed:
Merge branch 'trunk' into 9138-hostname-caching-policy
